components/brave_news/browser/feed_fetcher.cc

```diff
@@ -152,47 +157,86 @@ void FeedFetcher::OnFetchFeedFetchedFeed(
-  std::size_t total_size = 0;
-  for (const auto& result : results) {
-    total_size += result.items.size();
-  }
-  VLOG(1) << "All feed item fetches done with item count: " << total_size;
-
-  ETags etags;
-  FeedItems feed;
-  feed.reserve(total_size);
-
-  // We want to deduplicate the feed, as the feeds for different regions **may**
-  // have overlap.
-  base::flat_set<GURL> seen;
-
-  for (auto& result : results) {
-    etags[result.key] = result.etag;
-    for (auto& item : result.items) {
-      GURL url;
-      if (item->is_article()) {
-        url = item->get_article()->data->url;
-      } else if (item->is_promoted_article()) {
-        url = item->get_promoted_article()->data->url;
-      }
-
-      // Skip this, we've already seen it.
-      if (!url.is_empty() && seen.contains(url)) {
-        continue;
-      }
-      seen.insert(url);
-
-      feed.push_back(std::move(item));
-    }
-  }
-
-  std::move(callback).Run(std::move(feed), std::move(etags));
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(
+          [](std::vector<FeedSourceResult> results) {
+            std::size_t total_size = 0;
+            for (const auto& result : results) {
+              total_size += result.items.size();
+            }
+            VLOG(1) << "All feed item fetches done with item count: "
+                    << total_size;
+
+            ETags etags;
+            FeedItems feed;
+            feed.reserve(total_size);
+
+            // We want to deduplicate the feed, as the feeds for different
+            // regions **may** have overlap.
+            std::unordered_set<std::string> seen;
+
+            // reserve |total_size| space in |seen|. This is more than we'll
+            // likely need but should be in the correct ballpark.
+            seen.reserve(total_size);
+
+            for (auto& result : results) {
+              etags[result.key] = result.etag;
+              for (auto& item : result.items) {
+                GURL url;
+                if (item->is_article()) {
+                  url = item->get_article()->data->url;
+                } else if (item->is_promoted_article()) {
+                  url = item->get_promoted_article()->data->url;
+                }
+
+                // Skip this, we've already seen it.
+                auto spec = url.spec();
+                if (!url.is_empty() && seen.contains(spec)) {
+                  continue;
+                }
+                seen.insert(std::move(spec));
+
+                feed.push_back(std::move(item));
+              }
+            }
+
+            return std::make_tuple(std::move(feed), std::move(etags));
+          },
+          std::move(results)),
+      base::BindOnce(
+          [](base::WeakPtr<FeedFetcher> fetcher, FetchFeedCallback callback,
+             std::tuple<FeedItems, ETags> result) {
+            // If we've been destroyed, don't run the callback.
+            if (!fetcher) {
```

Comment:
Checking `fetcher` makes sense. Although, the lifetime of the callback could be different.
Frankly, I can't 100% sure that we don't add new crashes because posting the callback.

The code use a very callback-based approach. It's really hard to surely find all the entry points and make a conclusion that it's still safe..

Reply:
I other words, we changed the public contract of the class. Before we run all the callback sync, now we could postpone. A callback could be bind not only to `FeedFetcher` but to mojo-related interface (to send the answer) or something else. Better to double check all the users. Or even make renaming `DoSomething(..)` => `DoSomethingAsync(..)`

Reply:
I don't think we have changed the the public contract of the class - the callback was never invoked synchronously, only ever after all the combined feeds and direct feeds are fetched (quite apart from the network request, JSON parsing of the combined feed potentially happens in another process).

I added the check that `fetcher` wasn't destroyed because it was the only place I could think of where the contract was changing (previously the callback wouldn't be invoked if the `FeedFetcher` was destroyed, and the WeakPtr check maintains that guarantee).